### PR TITLE
Stateful vs Stateless Architectures

### LINKED ISSUE #2966: Statefull vs. Stateless Architectures
### Proposed title of the article
Statefull vs. Stateless Architectures

### Introduction paragraph (2-3 paragraphs):
Networks protocols control how data is transferred between different devices in a network. However, you should understand that networking protocols are divided into two: stateful and stateless protocols. Additionally, whether something is stateful or stateless depends on how long the state of interaction with it is being recorded and how that information needs to be stored. 

### Key takeaways: 
- Definition and basic history of stateful and stateless architectures
- The various protocols that utilize the two architectures
- The differences between the two architectures 
- Some of the advantages of each of the architectures 

### References: 
N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Su-sid**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #2966

### CODE REVIEW

**Comment by Su-sid** [comment]
@Dawe-7 kindly continue on with the review.

**Comment by Dawe-7** [comment]
@Su-sid, someone else from the team will be handling your article.

**Comment by WanjaMIKE** [comment]
@EKahuha could you review this article. Thanks

**Comment by EKahuha** [comment]
@Su-sid I will be reviewing this article.

**Comment by lin-chez** [comment]
It will be my pleasure @EKahuha

**Comment by lin-chez** [comment]
It will be a pleasure @EKahuha

**Comment by Su-sid** [comment]
thanks for the confirmation @EKahuha  it will be a pleasure.

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+1.	You should know internet protocols such as HTTP, UDP, TCP DNS SMTP, FTP, TCP, POP3 and TELNET as I will only feature a basic discussion of these protocols within the article.
+2.	Have a basic understanding of the difference between Graph QL, REST and SOAP application programming interfaces (APIs) 
+3.	You should have founding knowledge of networks bandwidths, servers, and clients.
+
+### Introduction 
+During the 1980s,  servers handled most the complex computations as clients were limited in their scope of operations that they could fulfill . However, as time passed, a global increase in web usage led to the development of the Hypertext Transfer Protocol (HTTP), web browsers, and other client-side software and hardware.
+
+Coincidentally, the rise in web usage led to an increase in the number of clients, as the clients could also handle many of the functions previously performed on servers.
+
+### What Are Stateful And Stateless Architectures
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:39]: "and" should be lowercase

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+3.	You should have founding knowledge of networks bandwidths, servers, and clients.
+
+### Introduction 
+During the 1980s,  servers handled most the complex computations as clients were limited in their scope of operations that they could fulfill . However, as time passed, a global increase in web usage led to the development of the Hypertext Transfer Protocol (HTTP), web browsers, and other client-side software and hardware.
+
+Coincidentally, the rise in web usage led to an increase in the number of clients, as the clients could also handle many of the functions previously performed on servers.
+
+### What Are Stateful And Stateless Architectures 
+
+These are the architectural frameworks that define how users interact with services on the internet. Stateful architecture-based services maintain track of clients records or sessions and apply different functionalities to similar user inputs based on the user's previous records. In a stateless design, services rely on the clients to keep track of transaction records, allowing them to concentrate on operations that change the system’s resources rather than its state.
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:41]: These are the architectural frameworks...Delete article 'the'

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+The main difference between the two architectures is how they handle states. In a stateful based system, additional information is reserved on the server-side, saving the present transaction's state as it waits for the proceeding user requests. 
+
+In a stateless build system, extra information is  held on the client-side. Thus, as the system conveys different information, each session'reminds' the server of its preceeding sessions through the use of Cookies.
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:45]: 'Cookies' should be lowercase

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+### Network Protocols and their Architectures
+
+Network protocols are divided into two: 
+- Stateful protocols 
+- Stateless protocols. 
+
+### Examples
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:53]: Restructure these subheadings to:  Examples of Stateless Protocols 
                                                          Examples of Stateful protocols

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+#### Statefull Protocols
+- The File Transfer Protocol (FTP).
+
+> The traditional FTP server couples with the client in an interactive session where while in session, a user is provided with a way to verify and set a few variables, all stored server-side, as part of the session state.
+
+- Telnet 
+- The Transmission Control Protocol (TCP)
+
+### API-based Internet Services and their Relation to Stateful and Stateless Architectures
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:74]: API-Based

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+The move to switch from a statefuk designed system to a stteless alternative is linked to the popularity of the diverse ways of building application programming interfaces (APIs) and web apps. At the start of the 21st century, the Simple Object Access Protocol (SOAP) was the prevalent design pattern. The design pattern allowed the creation of stateful architectures, evident from the services and applications that were develooed during the period.
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:76]: Do you mean 'statefuk' or 'stateful'?
 'stteless' or 'stateless' 
'develooed' or 'developed'

..... was a prevalent design pattern....

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+> The traditional FTP server couples with the client in an interactive session where while in session, a user is provided with a way to verify and set a few variables, all stored server-side, as part of the session state.
+
+- Telnet 
+- The Transmission Control Protocol (TCP)
+
+### API-based Internet Services and their Relation to Stateful and Stateless Architectures
+
+The move to switch from a statefuk designed system to a stteless alternative is linked to the popularity of the diverse ways of building application programming interfaces (APIs) and web apps. At the start of the 21st century, the Simple Object Access Protocol (SOAP) was the prevalent design pattern. The design pattern allowed the creation of stateful architectures, evident from the services and applications that were develooed during the period. 
+
+RESTful services developed in popularity as a result of its stateless design, eventually surpassing SOAP in usage. In the design world, Graph QL is gaining traction. RESTful services, like Graph QL-based services, are stateless thus e nd-users have a lot of influence over what data is sent to and from the servers.
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:78]: 'e nd-users' or 'end-users'?

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+### Differences Between Atateless and Atateful Architectures
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:80]: Atateless and Atateful?

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+The main difference between the two architectures is how they handle user states and sessions. The differences between the two areas are highlighted in the table below:
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:81]: The main differences between the two architectures relate to how they handle user states and sessions. These differences are highlighted in the table below.

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+**Stateless**|**Stateful**
+:-----:|:-----:
+Rely on clients to store session state information using cookies and local storage.|Rely on servers to store session state information.
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:85]: Relies

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+Server design is simple as it does not handle most of the functionality.|Server design is complex as it handles a lot of complexity
+The server handles requests much faster as it does not need to look beyond the single request sent to it as the request contains all the information the server needs to handle it.|The server handles user requests relatively slower compared to stateless architectures servers. The server has to manage the state sessions all on its own, thus, a response gets to the client much slower.
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:87]: Break this sentence 'The server handles requests much faster as it does not need to look beyond the single request sent to it as the request contains all the information the server needs to handle it., into two sentences.

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+The server and client are independent and loosely coupled.|The server and client are dependent on each other.
+Servers in this architecture are crash-proof. They can be restarted after a failure. The other servers could continue to handle a user request when the server handling it goes down.|Servers in this architecture do not recover easily from a crash. They store session status and information about a client during sessions. So if they crash, the session has to start over.
+
+
+### Advantages Stateless Architecture
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:92]: Advantages of Stateless Architectures

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+1.	A network's monitoring system does not need to read the information beyond a single request to comprehend it, which improves visibility. This is because, every request is handled in isolation.
+2.	They simplify the design of the server since most functionality is handled client-side.
+3.	It is very reliable as it recovers from partial failures easily. The server does not need to store any particular state or session; thus, a failed server can restart or transfer the client's request to another server of the same organization.
+4.	Scalability is achieved easily. Since the servers do not have to store session states, resources can be freed to handle other incoming requests.
+5.	They can easily handle multiple sessions simultaneously since every request is isolated.
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:97]: They can handle multiple sessions .....

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+6.	Memory usage on the server-side is reduced. This is because the information used to handle requests is not stored there.
+
+### Disadvantages of Stateless Protocols
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:100]: The Disadvantage of Stateless Protocols. 
You've only provided one disadvantage, so it cannot be disadvantages. 

Also this 'Advantages of Stateful Architecture'

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+3.	It is very reliable as it recovers from partial failures easily. The server does not need to store any particular state or session; thus, a failed server can restart or transfer the client's request to another server of the same organization.
+4.	Scalability is achieved easily. Since the servers do not have to store session states, resources can be freed to handle other incoming requests.
+5.	They can easily handle multiple sessions simultaneously since every request is isolated.
+6.	Memory usage on the server-side is reduced. This is because the information used to handle requests is not stored there.
+
+### Disadvantages of Stateless Protocols 
+1.	Recurring poor network performances. This might occur when increasing the redundant data sent in sequences of requests. Remember, we mentioned that data is not left on the server in a stateless network architecture.
+ ### Advantages of Stateful Architecture
+1.	Stateful architecture may add a security advantage to systems. That is why many banks rely on it to allow users to make online transactions. Its disadvantages, however, outweigh its benefits.
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:103]: A stateful

File: content/articles/statefull-vs-stateless-architectures/index.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+### Disadvantages of Stateful Architecture include:
+1.	It does not easily recover from partial failures. When a server fails, it cannot just restart or pass a user request to another server because it needs to hold the session state along with the details.
+2.	As much of the complex functionality is handled server-side, an intricate server design is needed to design the server.
+3.	Too much dependency between server and client that when one side fails, all connections are lost, and the running process has to start over.
+
+### Conclusion
+
+Majority of present systmes are designed using a stateless design. The statelsss design gained popularity from the server focused statefull way where clients were passive interfaces dependent on servers to perform majority of the computational needs. Developments in technology coupled with introduction of powerfull client computers allowed the shift to a stateless preferred environment that adds time to focus on the product development than on technical issues like server software implementation frequently encounetered in a statefull design. Stateless systems also enjoy the ability to scale in case a product or service explodes in a market.
```
Comment (EKahuha) [inline content/articles/statefull-vs-stateless-architectures/index.md:111]: The sentences in this paragraph are very long. Shorten them for ease of readability.

**Review by EKahuha — CHANGES_REQUESTED** [review]

**Comment by EKahuha** [comment]
Great article @Su-sid. 
There is an error in your article title. Use the correct spelling of stateful. 
Kindly ensure to proofread your article or use a grammar editor like Grammarly. 
Implement the suggested changes so that we can move on to tech review.

**Review by Su-sid — COMMENTED** [review]
All changes have been observed and implemented. Kindly continue on with the revision process @EKahuha

**Comment by EKahuha** [comment]
Also, correct the spelling of 'stateful' on your article and image paths.

**Comment by Su-sid** [comment]
Changes completed. Any other area that needs my attention?

**Comment by EKahuha** [comment]
> Also, correct the spelling of 'stateful' on your article and image paths.

content/articles/**statefull**-vs-stateless-architectures/index.md
content/articles/**statefull**-vs-stateless-architectures/hero.png

**Comment by Su-sid** [comment]
Done

**Comment by hectorkambow** [comment]
@Su-sid 
Please be sure you are going over the entire article as our peer reviewers are not responsible to point out every mistake - that is the authors responsibility.

cc @EKahuha

**Comment by hectorkambow** [comment]
@EKahuha 

FYI your review is too extensive - we should not be fixing the articles for the authors - our peer reviewer team is here  to review not revise. 

A PR of this quality should be close to allow the author the proper time to look over their work.

**Comment by Su-sid** [comment]
@EKahuha minor changes have been made to the pr. Kindly approve them for the article to continue on to its final stages. Kindly add your information to the `Peer Review Contributions` section.

**Comment by EKahuha** [comment]
@Su-sid I am waiting for a plagiarism check so that we can move to final review.

**Comment by Su-sid** [comment]
Kindly, can you do a plagiarism check on this PR?  @WanjaMIKE ,@Kayere or @LinusMuema

**Comment by Kayere** [comment]
Plagiarism check done.

**Comment by WanjaMIKE** [comment]
@EKahuha can I add the `ready for final review label`?

**Review by EKahuha — CHANGES_REQUESTED** [review]
@Su-sid 
I have reviewed your article's first paragraph, and I can still find grammatical errors. Run your article through Grammarly editor and proofread it thoroughly.

File: content/articles/stateful-vs-stateless-architectures/index.md [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+date: 2021-11-26T00:00:00-09:00
+topics: []
+excerpt_separator: <!--more-->
+images:
+
+  - url: /stateful-vs-stateless-architectures/hero.png
+    alt: The Stateful vs Stateless Architectures Hero Image
+---
+
+When developing applications and services, developers consider several factors to ensure their applications run in the most optimum way. Factors such as the scalability of the application/service, the hardware need such as networks speed and bandwidth required by the application, and the amount of temporary and permanent storage form a few of the factors that developers consider during the application development process. Additionally, developers consider the processing time, mainly determined by processor speed and application building architecture.
```
Comment (EKahuha) [inline content/articles/stateful-vs-stateless-architectures/index.md:18]: optimal way
hardware needs

**Comment by EKahuha** [comment]
> @EKahuha can I add the `ready for final review label`?

Not yet @WanjaMIKE

**Review by Su-sid — COMMENTED** [review]
.

**Comment by EKahuha** [comment]
@Su-sid  I sent you a messsage on slack

**Comment by EKahuha** [comment]
This is ready for final review @WanjaMIKE

**Comment by WanjaMIKE** [comment]
Nice test review @EKahuha

**Comment by WanjaMIKE** [comment]
Merged #5284 into master

**Pull request merged**